# Capstone Project - The Battle of the Neighborhoods (Week 2)

### Applied Data Science Capstone by IBM/Coursera

## Table of contents

* Introduction: Business Problem
* Data
* Methodology
* Analysis
* Results and Discussion
* Conclusion

## Introduction: Business Problem

New York City revolves around coffee. Considering there are 3,389 coffee shops in NYC. New York was one of the largest coffee roasting centers in the United States. More recently, mobile startup MassiveHealth found that New Yorkers drink 6.7 times the amount of coffee consumed by the average denizen of any other US city. And because New York loves Coffee so much, this makes it one of the best place to open a coffee shop.


We want to open a coffee shop in the New York City New York. i want to see what area has the most foot tracffic and to see if my coffee business will be profitable.

## Data

#### Based on definition of our problem, factors that will influence our decission are:

* number of existing coffee shot in the neighborhood (any type of Coffee shop)
* number of and distance to Coffee Shops in the neighborhood, if any
* distance of neighborhood from city center


We decided to use regularly spaced grid of locations, centered around city center, to define our neighborhoods.

Following data sources will be needed to extract/generate the required information:

* centers of candidate areas will be generated algorithmically and approximate addresses of centers of those areas will be obtained using Google Maps API reverse geocoding
* number of Coffee Shop and their type and location in every neighborhood will be obtained using Foursquare API
* we will also going to Run k-means to cluster the neighborhood into 5 clusters

## Neighborhood Candidates

Let's create latitude & longitude coordinates for centroids of our candidate neighborhoods. We will create a grid of cells covering our area of interest which is aprox. 12x12 killometers centered around New York City center.

Let's import all the Librarys we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
print('Libraries imported.')

Libraries imported.


In [2]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Folium installed and imported!')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.2         |   py36h9f0ad1d_0         152 KB  conda-forge
    ca-certificates-2020.4.5.2 |       hecda079_0         147 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                       

In [3]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [4]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [5]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

Notice how all the relevant data is in the features key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [6]:
neighborhoods_data = newyork_data['features']

Let's take a look at the first item in this list.

In [7]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

## Tranform the data into a pandas dataframe

The next task is essentially transforming this data of nested Python dictionaries into a pandas dataframe. So let's start by creating an empty dataframe.

In [8]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Take a look at the empty dataframe to confirm that the columns are as intended.

In [9]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [10]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [11]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [12]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


## Use geopy library to get the latitude and longitude values of New York City

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent ny_explorer, as shown below.

In [13]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


## Create a map of New York with neighborhoods superimposed on top.

In [14]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [15]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

In [16]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.

In [17]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

## Define Foursquare Credentials and Version

In [18]:
CLIENT_ID = 'BDYVYYS5YE3WRMWF23UOYAHF2MQ35N5YTUFOC4AIIWZ1EDXZ' # your Foursquare ID
CLIENT_SECRET = 'N42J5Y2OBOJRAJF24UN150TAB0YPKFSCPPPNLIMNQEYA0QLP' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: BDYVYYS5YE3WRMWF23UOYAHF2MQ35N5YTUFOC4AIIWZ1EDXZ
CLIENT_SECRET:N42J5Y2OBOJRAJF24UN150TAB0YPKFSCPPPNLIMNQEYA0QLP


## Let's explore the first neighborhood in our dataframe

Get the neighborhood's name.

In [19]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

Get the neighborhood's latitude and longitude values.

In [20]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


In [21]:
 LIMIT = 100 
radius = 500      
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=BDYVYYS5YE3WRMWF23UOYAHF2MQ35N5YTUFOC4AIIWZ1EDXZ&client_secret=N42J5Y2OBOJRAJF24UN150TAB0YPKFSCPPPNLIMNQEYA0QLP&v=20180605&ll=40.87655077879964,-73.91065965862981&radius=500&limit=100'

In [22]:
results = requests.get('https://api.foursquare.com/v2/venues/explore?&client_id=BDYVYYS5YE3WRMWF23UOYAHF2MQ35N5YTUFOC4AIIWZ1EDXZ&client_secret=N42J5Y2OBOJRAJF24UN150TAB0YPKFSCPPPNLIMNQEYA0QLP&v=20180605&ll=40.87655077879964,-73.91065965862981&radius=500&limit=100').json()
results

{'meta': {'code': 200, 'requestId': '5eeac15e6d8c560028d2367b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 25,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b4429abf964a52037f225e3',
       'name': "Arturo's",
       'location': {'address': '5198 Broadway',
        'crossStreet': 'at 225th St.',
        'lat': 40.87441177110231,
        'lng': -73.91027100981574,
        'labeledLatLngs': [{'label'

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function from the Foursquare lab.

In [23]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [24]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name   categories        lat        lng
0       Arturo's  Pizza Place  40.874412 -73.910271
1    Bikram Yoga  Yoga Studio  40.876844 -73.906204
2  Tibbett Diner        Diner  40.880404 -73.908937
3      Starbucks  Coffee Shop  40.877531 -73.905582
4        Dunkin'   Donut Shop  40.877136 -73.906666

And how many venues were returned by Foursquare?

In [25]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

25 venues were returned by Foursquare.


## Explore Neighborhoods in Manhattan

Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Now write the code to run the above function on each neighborhood and create a new dataframe called manhattan_venues

In [28]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


### Let's check the size of the resulting dataframe

In [29]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3148, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066      Starbucks   
4  Marble Hill              40.876551               -73.91066        Dunkin'   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877531       -73.905582    Coffee Shop  
4       40.877136       -73.906666     Donut Shop

Let's check how many venues were returned for each neighborhood

In [30]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 329 uniques categories.


## Analyze Each Neighborhood

In [31]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Baby Store  \
0                      0                    0          0           0   
1                      0                    0          0           0   
2                      0                    0          0           0   
3                      0                    0          0           0   
4                      0                    0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Stadium  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                   0         0            0           0              0   
1                   0         0            0           0              0   
2                   0         0            0           0              0   
3                   0         0            0           0              0   
4                   0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0              0          0         0           0                     0   
1              0          0         0           0                     0   
2              0          0         0           0                     0   
3              0          0         0           0                     0   
4              0          0         0           0                     0   

   Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  Building  \


And let's examine the new dataframe size.

In [32]:
manhattan_onehot.shape

(3148, 330)

### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [33]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.000000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.000000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.00               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.000000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.000000            0.00               0.00   
37   Washington Heights           0.011364            0.00               0.00   
38         West Village           0.000000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Antique Shop    Arcade  \
0             0.000000             0.015152          0.00  0.000000   
1             0.000000             0.011236          0.00  0.000000   
2             0.041667             0.041667          0.00  0.000000   
3             0.000000             0.040000          0.00  0.000000   
4             0.000000             0.030000          0.00  0.000000   
5             0.000000             0.030000          0.01  0.000000   
6             0.000000             0.030000          0.00  0.000000   
7             0.000000             0.000000          0.00  0.000000   
8             0.000000             0.010000       

Let's confirm the new size

In [34]:
manhattan_grouped.shape

(40, 330)

### Let's print each neighborhood along with the top 5 most common venues

In [35]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
           venue  freq
0           Park  0.12
1          Hotel  0.06
2    Coffee Shop  0.06
3  Boat or Ferry  0.05
4  Memorial Site  0.05


----Carnegie Hill----
                venue  freq
0         Coffee Shop  0.09
1                Café  0.04
2  Italian Restaurant  0.04
3         Pizza Place  0.04
4                 Gym  0.03


----Central Harlem----
                  venue  freq
0    Chinese Restaurant  0.06
1  Gym / Fitness Center  0.04
2    African Restaurant  0.04
3   American Restaurant  0.04
4           Pizza Place  0.04


----Chelsea----
                 venue  freq
0          Coffee Shop  0.08
1          Art Gallery  0.07
2       Ice Cream Shop  0.04
3  American Restaurant  0.04
4    French Restaurant  0.03


----Chinatown----
                venue  freq
0  Chinese Restaurant  0.08
1              Bakery  0.07
2        Dessert Shop  0.04
3     Bubble Tea Shop  0.03
4                 Bar  0.03


----Civic Center----
                  venue  freq
0     

### Let's put that into a pandas dataframe


First, let's write a function to sort the venues in descending order.

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park                 Hotel   
1      Carnegie Hill           Coffee Shop                  Café   
2     Central Harlem    Chinese Restaurant    Seafood Restaurant   
3            Chelsea           Coffee Shop           Art Gallery   
4          Chinatown    Chinese Restaurant                Bakery   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop          Memorial Site                   Gym   
1           Pizza Place     Italian Restaurant           Yoga Studio   
2           Pizza Place                    Bar     French Restaurant   
3        Ice Cream Shop    American Restaurant     French Restaurant   
4          Dessert Shop  Vietnamese Restaurant          Cocktail Bar   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Boat or Ferry            Food Court         Shopping Mall   
1             Bookstore             Wine Shop  Gym / Fitness Center   
2  Gym / Fitness Center   American Restaurant    African Restaurant   
3                Bakery                  Café          Cocktail Bar   
4                   Bar                   Spa        Ice Cream Shop   

  9th Most Common Venue 10th Most Common Venue  
0          Gourmet Shop     Mexican Restaurant  
1                   Gym      French Restaurant  
2         Deli / Bodega              Bookstore  
3           Pizza Place                 Market  
4       Bubble Tea Shop      Hotpot Restaurant

## Cluster Neighborhoods


Run k-means to cluster the neighborhood into 5 clusters.

In [38]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 0, 2, 1, 0, 1, 1, 2, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [39]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               2   
1  Manhattan           Chinatown  40.715618 -73.994279               0   
2  Manhattan  Washington Heights  40.851903 -73.936900               2   
3  Manhattan              Inwood  40.867684 -73.921210               2   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               2   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Sandwich Place           Coffee Shop                   Gym   
1    Chinese Restaurant                Bakery          Dessert Shop   
2                  Café                Bakery           Pizza Place   
3    Mexican Restaurant                Lounge                  Café   
4           Pizza Place         Deli / Bodega           Coffee Shop   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0               Pharmacy      Department Store                 Diner   
1  Vietnamese Restaurant          Cocktail Bar                   Bar   
2          Grocery Store     Mobile Phone Shop            Donut Shop   
3             Restaurant                  Park    Frozen Yogurt Shop   
4                   Café    Mexican Restaurant                Bakery   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Discount Store            Donut Shop            Kids Store   
1                   Spa        Ice Cream Shop       Bubble Tea Shop   
2    Mexican Restaurant        Sandwich Place                  Bank   
3                Bakery    Spanish Restaurant    Chinese Restaurant   
4      Sushi Restaurant          Cocktail Bar        Sandwich Place   

  10th Most Common Venue  
0          Deli / Bodega  
1      Hotpot Restaurant  
2            Coffee Shop  
3    American Restaurant  
4            Yoga Studio

Finally, let's visualize the resulting clusters

In [40]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters


Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

### Cluster 1

In [41]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1            Chinatown    Chinese Restaurant                Bakery   
12     Upper West Side    Italian Restaurant                   Bar   
19        East Village                   Bar    Mexican Restaurant   
25    Manhattan Valley           Coffee Shop                   Bar   
27            Gramercy                   Bar            Bagel Shop   
29  Financial District           Coffee Shop                   Bar   
30       Carnegie Hill           Coffee Shop                  Café   
31                Noho           Pizza Place    Italian Restaurant   

   3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
1           Dessert Shop  Vietnamese Restaurant          Cocktail Bar   
12          Dessert Shop                 Bakery              Wine Bar   
19          Cocktail Bar      Korean Restaurant           Pizza Place   
25    Mexican Restaurant            Pizza Place           Yoga Studio   
27           Pizza Place            Coffee Shop   American Restaurant   
29   American Restaurant           Cocktail Bar           Pizza Place   
30           Pizza Place     Italian Restaurant           Yoga Studio   
31         Grocery Store            Coffee Shop   Japanese Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1                    Bar                   Spa        Ice Cream Shop   
12     Indian Restaurant           Coffee Shop       Thai Restaurant   
19           Coffee Shop              Wine Bar        Ice Cream Shop   
25     Health Food Store   Peruvian Restaurant                   Spa   
27    Italian Restaurant            Playground         Grocery Store   
29    Italian Restaurant             Juice Bar            Steakhouse   
30             Bookstore             Wine Shop  Gym / Fitness Center   
31          Cocktail Bar     French Restaurant                 Hotel   

   9th Most Common Venue 10th Most Common Venue  
1        Bubble Tea Shop      Hotpot Restaurant  
12        Breakfast Spot         Ice Cream Shop  
19   Japanese Restaurant              Speakeasy  
25       Bubble Tea Shop                   Café  
27          Cocktail Bar     Mexican Restaurant  
29                  Park            Event Space  
30                   Gym      French Restaurant  
31    Mexican Restaurant               Wine Bar

### Cluster 2

In [42]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
13     Lincoln Square                  Café                 Plaza   
14            Clinton               Theater    Italian Restaurant   
15            Midtown           Coffee Shop                 Hotel   
16        Murray Hill        Sandwich Place           Coffee Shop   
17            Chelsea           Coffee Shop           Art Gallery   
28  Battery Park City                  Park                 Hotel   
32       Civic Center           Coffee Shop  Gym / Fitness Center   
33      Midtown South     Korean Restaurant                 Hotel   
38           Flatiron  Gym / Fitness Center                  Café   
39       Hudson Yards                 Hotel  Gym / Fitness Center   

       3rd Most Common Venue    4th Most Common Venue  5th Most Common Venue  \
13        Italian Restaurant                  Theater  Performing Arts Venue   
14      Gym / Fitness Center              Coffee Shop              Wine Shop   
15                    Bakery                  Theater         Clothing Store   
16       Japanese Restaurant                    Hotel            Pizza Place   
17            Ice Cream Shop      American Restaurant      French Restaurant   
28               Coffee Shop            Memorial Site                    Gym   
32              Cocktail Bar                      Spa                  Hotel   
33                      Café      Japanese Restaurant                 Lounge   
38  Mediterranean Restaurant  New American Restaurant            Coffee Shop   
39       American Restaurant       Italian Restaurant                    Gym   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
13          Concert Hall  Gym / Fitness Center           Coffee Shop   
14                   Gym        Sandwich Place          Cocktail Bar   
15           Pizza Place      Cuban Restaurant            Steakhouse   
16   American Restaurant  Gym / Fitness Center    Chinese Restaurant   
17                Bakery                  Café          Cocktail Bar   
28         Boat or Ferry            Food Court         Shopping Mall   
32     French Restaurant           Yoga Studio                  Café   
33          Burger Joint      Cuban Restaurant           Coffee Shop   
38                  Park    Italian Restaurant                   Gym   
39                   Bar                  Park               Dog Run   

   9th Most Common Venue         10th Most Common Venue  
13             Wine Shop            American Restaurant  
14   American Restaurant                          Hotel  
15        Sandwich Place                      Bookstore  
16                  Café                      Juice Bar  
17           Pizza Place                         Market  
28          Gourmet Shop             Mexican Restaurant  
32                  Park            American Restaurant  
33          Cocktail Bar                   Dessert Shop  
38   Japanese Restaurant  Vegetarian / Vegan Restaurant  
39           Coffee Shop                     Restaurant

### Cluster 3

In [43]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  2nd Most Common Venue  \
0           Marble Hill        Sandwich Place            Coffee Shop   
2    Washington Heights                  Café                 Bakery   
3                Inwood    Mexican Restaurant                 Lounge   
4      Hamilton Heights           Pizza Place          Deli / Bodega   
5        Manhattanville         Deli / Bodega            Coffee Shop   
6        Central Harlem    Chinese Restaurant     Seafood Restaurant   
7           East Harlem    Mexican Restaurant        Thai Restaurant   
11     Roosevelt Island                  Park  Outdoors & Recreation   
20      Lower East Side    Chinese Restaurant               Pharmacy   
26  Morningside Heights             Bookstore    American Restaurant   
36           Tudor City                  Café                   Park   

                       3rd Most Common Venue 4th Most Common Venue  \
0                                        Gym              Pharmacy   
2                                Pizza Place         Grocery Store   
3                                       Café            Restaurant   
4                                Coffee Shop                  Café   
5                         Mexican Restaurant    Seafood Restaurant   
6                                Pizza Place                   Bar   
7                                     Bakery         Deli / Bodega   
11  Residential Building (Apartment / Condo)                School   
20                                      Park                Bakery   
26                                      Park           Coffee Shop   
36                        Mexican Restaurant                 Diner   

   5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0       Department Store                      Diner        Discount Store   
2      Mobile Phone Shop                 Donut Shop    Mexican Restaurant   
3                   Park         Frozen Yogurt Shop                Bakery   
4     Mexican Restaurant                     Bakery      Sushi Restaurant   
5     Italian Restaurant         Chinese Restaurant   Dumpling Restaurant   
6      French Restaurant       Gym / Fitness Center   American Restaurant   
7         Sandwich Place  Latin American Restaurant           Gas Station   
11          Liquor Store             Sandwich Place              Bus Line   
20                  Café               Cocktail Bar           Art Gallery   
26        Sandwich Place              Deli / Bodega          Burger Joint   
36         Deli / Bodega                Pizza Place               Dog Run   

      8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
0                Donut Shop                Kids Store          Deli / Bodega  
2            Sandwich Place                      Bank            Coffee Shop  
3        Spanish Restaurant        Chinese Restaurant    American Restaurant  
4              Cocktail Bar            Sandwich Place            Yoga Studio  
5                      Park          Sushi Restaurant            Supermarket  
6        African Restaurant             Deli / Bodega              Bookstore  
7              Liquor Store                Steakhouse          Grocery Store  
11              Coffee Shop               Dry Cleaner       Greek Restaurant  
20         Ramen Restaurant               Coffee Shop  Performing Arts Venue  
26  New American Restaurant  Mediterranean Restaurant     Mexican Restaurant  
36              Coffee Shop                    Garden     Seafood Restaurant

### Cluster 4

In [44]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
8     Upper East Side    Italian Restaurant              Coffee Shop   
9           Yorkville    Italian Restaurant                      Gym   
10         Lenox Hill    Italian Restaurant              Coffee Shop   
18  Greenwich Village    Italian Restaurant                     Café   
21            Tribeca    Italian Restaurant                     Park   
22       Little Italy                Bakery       Chinese Restaurant   
23               Soho    Italian Restaurant           Clothing Store   
24       West Village    Italian Restaurant  New American Restaurant   
34       Sutton Place  Gym / Fitness Center       Italian Restaurant   
35         Turtle Bay    Italian Restaurant              Coffee Shop   

       3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
8                     Bakery  Gym / Fitness Center                 Juice Bar   
9                Coffee Shop      Sushi Restaurant                       Bar   
10               Pizza Place                  Café              Cocktail Bar   
18          Sushi Restaurant        Clothing Store         French Restaurant   
21       American Restaurant                   Spa                  Wine Bar   
22                       Spa       Bubble Tea Shop  Mediterranean Restaurant   
23  Mediterranean Restaurant        Sandwich Place               Coffee Shop   
24              Cocktail Bar                  Park                  Wine Bar   
34               Coffee Shop                  Park    Furniture / Home Store   
35                      Park                  Café          Sushi Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
8            Yoga Studio     French Restaurant                   Spa   
9          Deli / Bodega           Pizza Place   Japanese Restaurant   
10      Sushi Restaurant  Gym / Fitness Center                   Gym   
18        Ice Cream Shop    Chinese Restaurant        Sandwich Place   
21                  Café      Greek Restaurant           Coffee Shop   
22    Italian Restaurant                  Café        Cosmetics Shop   
23     French Restaurant                   Spa                  Café   
24   American Restaurant           Pizza Place           Coffee Shop   
34            Bagel Shop                   Gym           Pizza Place   
35     French Restaurant         Deli / Bodega    Seafood Restaurant   

   9th Most Common Venue 10th Most Common Venue  
8          Women's Store       Sushi Restaurant  
9     Mexican Restaurant              Wine Shop  
10          Burger Joint        Thai Restaurant  
18        Pilates Studio                 Bakery  
21                Bakery                  Hotel  
22       Thai Restaurant         Sandwich Place  
23                 Hotel                 Bakery  
24                Bakery              Jazz Club  
34                   Spa         Lingerie Store  
35                   Bar    Japanese Restaurant

### Cluster 5

In [45]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                  Park        Baseball Field   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37              Heliport                   Bar          Cocktail Bar   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37           Coffee Shop  Gym / Fitness Center       Harbor / Marina   

   9th Most Common Venue 10th Most Common Venue  
37                Bistro            Gas Station

# Results and Discussion

Beased on our reseach, we could see that New York City have lots of coffee shop. As a matter of fact their are over 3,389 coffee shops in NYC. After looking at all the factors, we beleive NYC is the perfect location for a new coffee shop. We took a look at all the Borough in NYC and figured out where is the best place for a new coffee shop. We look at foot tracffic, other kinds of food stores around the real. After all the research, we believe Manhattan is the best location,

We than broke Manhattan down to Neighborhood to see where is the best place for a new coffee shop. we found out that Marble Hill is the best location for a new coffee shop. accounding to foursqure data, coffee shop is the 2nd most popular venue in Marble Hill next to sandwhich store. We also ran k-means to cluster the neighborhood into 5 clusters and in the secound cluster, we could clearly see that coffee shop is very popular.

# Conclusion

# Why coffee shop


Coffee shops are undoubtedly a small piece of the much larger puzzle that is contemporary urban living; yet their importance should not be underestimated. From their influence on the enlightenment to their ever-evolving status. Openning a  coffee shop is my own small way of contributing to my community and the culture they participate in, create, and project — have played a vital, often controversial role in shaping urban experience. Thus understanding coffee shops in a variety of ways